crates/red_knot_python_semantic/src/types/call.rs

```diff
@@ -249,6 +258,54 @@ impl<'db> CallOutcome<'db> {
                     }),
                 }
             }
+            CallOutcome::StaticAssertionError {
+                parameter_ty: Type::BooleanLiteral(false),
+                truthiness: _,
+            } => {
+                context.report_lint(
+                    &STATIC_ASSERT_ERROR,
+                    node,
+                    format_args!("Static assertion error: argument evaluates to `False`"),
+                );
+
+                Ok(Type::Unknown)
+            }
+            CallOutcome::StaticAssertionError {
+                parameter_ty,
+                truthiness,
+            } if truthiness.is_always_false() => {
+                context.report_lint(
+                            &STATIC_ASSERT_ERROR,
+                            node,
+                            format_args!("Static assertion error: argument of type `{parameter_ty}` is statically known to be falsy", parameter_ty=parameter_ty.display(db)),
+                        );
+
+                Ok(Type::Unknown)
+            }
+            CallOutcome::StaticAssertionError {
+                parameter_ty,
+                truthiness,
+            } if truthiness.is_ambiguous() => {
+                context.report_lint(
+                            &STATIC_ASSERT_ERROR,
+                            node,
+                            format_args!("Static assertion error: argument of type `{parameter_ty}` has an ambiguous static truthiness", parameter_ty=parameter_ty.display(db)),
+                        );
+
+                Ok(Type::Unknown)
+            }
+            CallOutcome::StaticAssertionError {
+                parameter_ty,
+                truthiness: _,
+            } => {
+                context.report_lint(
+                            &STATIC_ASSERT_ERROR,
+                            node,
+                            format_args!("Static assertion error: expected argument of type `{parameter_ty}` to have static truthiness", parameter_ty=parameter_ty.display(db)),
+                        );
+
+                Ok(Type::Unknown)
+            }
```

Comment:
Shouldn't this case be unreachable, because `truthiness` must be either always-false, ambiguous, or always-true, and we won't create a `CallOutcome::StaticAssertionError` if its always-true? I would be inclined to explicitly mark it as unreachable.

Reply:
I added a new `StaticAssertionErrorKind` enum that describes the various cases. This allows us to remove this case without having to use `unreachable!`.